Fix adjust remaining time with playback speed, #3814

low-batt opened the pull request:
The commit in the pull request will:
- Change DurationDisplayTextField.updateText to take a playSpeed
  parameter
- Change updateText to adjust the remaining time based on the speed
- Change PlayerWindowController.updatePlayTime to pass the playback
  speed to updateText

- [ ] This change has been discussed with the author.
- [x] It implements / fixes issue #3814.

---

**Description:**

low-batt committed:
Fix adjust remaining time with playback speed, #3814

The commit in the pull request will:
- Change DurationDisplayTextField.updateText to take a playSpeed
  parameter
- Change updateText to adjust the remaining time based on the speed
- Change PlayerWindowController.updatePlayTime to pass the playback
  speed to updateText

low-batt commented:
@specter78 Gives this PR a **thumbs down**, requesting that the user be able to choose if the remaining time reflects playback speed or not.

At the moment we are holding off creating PRs that change preferences to avoid lots of merge conflicts. Will revise this PR once the outstanding preference related changes have been merged.

specter78 commented:
This PR creates a change that is not preferred by me and maybe many others too. So, it should not be merged, till it provides a preference option to disable it.

low-batt commented:
Labeled with `other: do not merge` to make certain this PR is not merged until it has been updated with a preference to control this behavior.

low-batt commented:
I was reminded that you can convert a PR to a draft to make it clear it is not ready to be merged. I do plan on coming back to this PR and adding a preference to control this behavior.

low-batt committed:
Close adjust remaining time with playback speed, #3814

This commit will:
- Add a new setting "Base remaining time on playback speed" to the UI
  settings tab
- Add a new preference key scaleRemainingTime to store the state of the
  new setting
- Add a new playSpeed property to DurationDisplayTextField
- Change DurationDisplayTextField.updateText to take a playSpeed
  parameter
- Change updateText to adjust the remaining time based on the speed if
  the new setting is enabled
- Change PlayerWindowController.updatePlayTime to pass the playback
  speed to updateText

The new setting is subordinate to the existing "Show remaining time
instead of total duration" an only enabled if that setting is enabled.
By default the new setting is not enabled to preserve the existing
behavior.

low-batt commented:
I rebased and resolved merge conflicts and then added a new setting `Base remaining time on playback speed` to allow the user to control whether the remaining time takes into account the playback speed or not:

![issue-3814](https://github.com/iina/iina/assets/86170219/6a93964c-789c-4ed3-9fbe-5fe9612794ee)

uiryuu commented:
mpv has a property "playtime-remaining` which calculates the remaining playtime for us:
https://mpv.io/manual/stable/#command-interface-playtime-remaining. I think we should use this property instead of calculating by ourselves.

low-batt commented:
I'm going to move this one to b2 and revisit it when we start working on that release.

low-batt committed:
Close adjust remaining time with playback speed, #3814

This commit will:
- Add a new setting "Base remaining time on playback speed" to the UI
  settings tab
- Add a new preference key scaleRemainingTime to store the state of the
  new setting
- Add a new videoRemaining property to PlaybackInfo
- Change PlayerCore.fileLoaded to set videoRemaining to the value of the
  mpv properties playtime-remaining/full or time-remaining/full based on
  the new scaleRemainingTime setting
- Add code to PlayerCore.syncUI to update videoRemaining
- Add a new remaining property to DurationDisplayTextField
- Change DurationDisplayTextField.updateText to take a remaining
  parameter
- Change DurationDisplayTextField to use the new remaining property when
  displaying the remaining playback time
- Change PlayerWindowController.updatePlayTime to pass the remaining
  playback time to DurationDisplayTextField.updateText

The new setting is subordinate to the existing "Show remaining time
instead of total duration" and only enabled if that setting is enabled.
By default the new setting is not enabled to preserve the existing
behavior.

low-batt committed:
Close adjust remaining time with playback speed, #3814

This commit will remove some unintended changes to
PrefUIViewController.xib introduced when resolving merge conflicts.

low-batt committed:
Close adjust remaining time with playback speed, #3814

This commit will remove some unintended changes to
PrefUIViewController.xib introduced when resolving merge conflicts.

low-batt commented:
@uiryuu I've rebased with develop and corrected merge conflicts and changed the code to use the mpv properties.

I do have one doubt. As seen here the new subordinate setting is disabled if `Show remaining time instead of total duration` is not enabled:

![settings](https://github.com/user-attachments/assets/33055c27-2773-49a0-aa8a-2b485678b825)

I'm thinking that is a mistake on my part. The `Show remaining time instead of total duration` just controls whether the OSC shows remaining time by default. You can always click on the time to switch between showing duration or remaining time. So I'm thinking the new setting should not be disabled if `Show remaining time instead of total duration` is disabled.

If you agree I will make that change.

low-batt committed:
Close adjust remaining time with playback speed, #3814

This commit will allow the user to change the new subordinate "Base
remaining time on playback speed" even when the "Show remaining time
instead of total duration" is disabled.

low-batt commented:
I went ahead and made the change to allow the `Base remaining time on playback speed` regardless of the `Show remaining time instead of total duration` setting.

uiryuu reviewed (approved):
LGTM. Needs @lhc70000's input.
